### PR TITLE
[feat] Duplicate plants warning + delete plants functionality

### LINKED ISSUE #118: Add warning to add details for adding duplicate plants
https://linear.app/trap-garden/issue/TG-71/prevent-duplicate-plants-from-being-added-to-my-garden

### PR DESCRIPTION
**Pull request by torinnayak1**
### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

When attempting to add plants that are already in their garden, shows a popup that there are plants to be added that are already in the garden (naming the plants), giving the user the option to proceed and add the plants again or press no and edit their details. Also added a delete button next to the plants in the review page of the add details

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."

image

Screenshot 2025-04-17 at 6 08 00 PM

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

I think there are probably more efficient ways to write some of the backend functions, please let me know. Edge cases could be when having a mixture of duplicates and non duplicates and then removing some and trying again (works on my end but I doubt I tested everything).

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

Maybe consolidating same plants with different seasons into the same card? Also, should they be redirected to plants page if they remove all plants from the add details?

### Online sources
[//]: # 'Copy links to any tutorials or documentation that was useful to you when working on this PR'

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @kylezryr

### CODE REVIEW

File: app/add-details/page.tsx [hunk 1]
```diff
@@ -24,30 +28,52 @@ function ReviewPlant({
...
-      </H4>
-      <ReviewGrid>
-        <P2 $fontWeight={500}>Date Planted</P2>
-        <P2>{dateAdded}</P2>
-        <P2 $fontWeight={500}>Planting Type</P2>
-        <P2>{plantingTypeLabels[plantingType]}</P2>
-      </ReviewGrid>
-    </Flex>
+    <>
+      {show ? (
```
Comment (kylezryr) [inline app/add-details/page.tsx]: instead of having an empty `<></>` as the else case, you can just do `{show && ( ... )}` to only show the first case if show is true. if its false, nothing will show up

File: app/add-details/page.tsx [hunk 2]
```diff
@@ -119,32 +145,67 @@ export default function Home() {
...
+            date_added: detail.date_added!,
+            planting_type: detail.planting_type!,
+          }))
+          .filter(plant => plant != undefined);
+        const firstPress = await insertUserPlants(
+          completedDetails,
+          confirm,
+          userId,
+        );
+        console.log(firstPress);
```
Comment (kylezryr) [inline app/add-details/page.tsx]: remove console log

File: app/add-details/page.tsx [hunk 3]
```diff
@@ -119,32 +145,67 @@ export default function Home() {
...
+        if (!confirm) {
+          if (firstPress?.length != 0) {
```
Comment (kylezryr) [inline app/add-details/page.tsx]: we should be able to condense these first two if statements into just one `if (!confirm && firstPress?.length != 0)` since there's no else case for the first if

File: app/add-details/page.tsx [hunk 4]
```diff
@@ -119,32 +145,67 @@ export default function Home() {
...
+            if (firstPress) {
```
Comment (kylezryr) [inline app/add-details/page.tsx]: is this second check necessary? wouldn't firstPress be true if the if statement above is true? if its because you're getting errors about firstPress not being null, you can add an exclamation mark after firstPress (`firstPress!`) where you need to to assert its not null

File: app/add-details/page.tsx [hunk 5]
```diff
@@ -221,6 +282,9 @@ export default function Home() {
                   plantName={plantsToAdd[index].plant_name}
                   plantingType={detail.planting_type!}
                   dateAdded={detail.date_added!}
+                  removeFunction={() => {
+                    delete details[index];
```
Comment (kylezryr) [inline app/add-details/page.tsx]: i think this isn't working, because if i remove a plant in the review page but then go back to the add details page, i get this error. i think we might need to re-set the `details` state when removeFunction is called to a new array with the index excluded perhaps

image

File: components/ConfirmationModal/index.tsx [hunk 6]
```diff
@@ -13,16 +13,22 @@ interface ConfirmationModalProps {
...
   onConfirm: () => void;
+  flip?: boolean | false;
 }
 
 export default function ConfirmationModal({
   isOpen,
   title,
   message,
+  leftText,
+  rightText,
```
Comment (kylezryr) [inline components/ConfirmationModal/index.tsx:28]: we can assign default values to these props since they're optional, namely `leftText = 'Cancel'` and `rightText = 'Exit'`, and same for flip being false by default. can we also rename `flip` to `flipButtons` to be clearer

File: app/add-details/page.tsx
Comment (kylezryr) [inline app/add-details/page.tsx:1]: we should add a useEffect here to check if there are no more plants in the add details everytime `details` is updated (e.g. if you're on the review screen and delete all the plants) (whenever `details` is an empty list). if its empty, route the user back to `/view-plants` since there's nothing left for them to add

File: app/add-details/page.tsx [hunk 7]
```diff
@@ -119,32 +145,67 @@ export default function Home() {
...
+            if (firstPress) {
+              const plantNames = await Promise.all(
+                firstPress.map(async dup => {
+                  const plant = await getPlantById(dup.plant_id);
+                  return plant?.plant_name;
+                }),
+              );
+
+              // Remove any undefined values (in case a lookup failed)
+              setDuplicates(
```
Comment (kylezryr) [inline app/add-details/page.tsx]: could we do this filtering to remove undefined values in the `insertUserPlants` function?

File: app/add-details/page.tsx [hunk 8]
```diff
@@ -24,30 +28,52 @@ function ReviewPlant({
...
   dateAdded,
   plantingType,
+  removeFunction,
 }: {
   plantName: string;
   dateAdded: string;
   plantingType: PlantingTypeEnum;
+  removeFunction: () => void;
 }) {
+  const [show, setShow] = useState(true);
```
Comment (kylezryr) [inline app/add-details/page.tsx]: nit: can we rename this to `showReviewRow` or smth coz I got confused with this and `showConfModal`

**Review by kylezryr — CHANGES_REQUESTED** [review]
works pretty well! just a bug with deleting a plant and then going back to the details page and its still there, resulting in an error. also to add a check to route the user to view plants if the details list is empty. good job overall!

**Review by kylezryr — APPROVED** [review]
ready 2 merge
